api/src/main/java/org/apache/iceberg/ChangelogScanTask.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+/**
+ * A changelog scan task.
+ */
+public interface ChangelogScanTask extends FileScanTask {
```

Comment:
I am debating whether we should extend `FileScanTask` or not.

Some benefits of extending:
- Our usage kind of fits the description: `A scan task over a range of a single file`.
- We can reuse split logic and `CombinedScanTask`.
- We can reuse the existing `planFiles` and `planTasks` methods.
- We can reuse a lot of reader code that assumes we get some sort of `FileScanTask`.

Reply:
Follow the discussing on https://github.com/apache/iceberg/pull/4812#discussion_r887384420, there's a discussion on whether we can and should refactor FileScanTask.  Today, this is baked into the code and used heavily:

1. Scan returns FileScanTask
2. FileScanTask has methods DataFile file() and List<DeleteFile> deleteFiles(), which makes sense only in case of scanning data files.
3. BaseFileScanTask has method Iterable<FileScanTask> split(), which uses file() to calculate the splits.

Problem: for growing number of cases , like CDC, read position deletes, writing a new scan implementation forces FileScanTask to be returned, which will then have to unnaturally wrap non-data files as DataFile file().  In these cases, even wrapping a DeleteFile as a DataFile.

I see two options

Option 1:  Leave it as is, implement various subclasses of DataFile that represent delete files, change logs, etc.  Minimal code change, but messy.  We will see DeleteDataFile.

Option 2:  Refactor FileScanTask hierarchy.
One potential solution is in #4812:

1. Scan still returns FileScanTask (cant change this)
2. FileScanTask now becomes a parameterized base class with method : ContentFile<T> contentFile()
3. BaseFileScanTask will do all the splitting logic based on contentFile().  In fact, this is a small code replacement of using file() => contentFile() as already today all the methods it uses are in fact from contentFile and not specific to DataFile.
4. Eventually, we make DataFileScanTask that has methods DataFile dataFile() and List<DeleteFile> deleteFiles(), and code that deals exclusively with DataFileScanTask can use that for clarity. 
5. Deprecate current FileScanTask file() and deleteFiles() and eventually remove them in favor of the subclass (though with 1.0, not sure the timeline we can do this).

With this solution, we need only minimal changes for 1-3, and can do more refactoring of 4 in different prs, while preserving backward compatibility.  I like this solution as it keeps some kind of clarity in the code that we are not dealing with DataFile but rather different content files.

But I am not entirely sure what kind of interface CDC FileScanTask will need, so will be open to see if this works.

Also, let me know if this is the not the place you meant for this discussion.

Reply:
> Option 1: Leave it as is, implement various subclasses of DataFile that represent delete files, change logs, etc. Minimal code change, but messy. We will see DeleteDataFile.

I don't think we will have to add `DeleteDataFile`. We can simply construct `DataFile` from `DeleteFile` like we construct `DataFile` from `ManifestFile` in metadata tables today.

```
DataFiles.fromManifest(manifest)
```

It is a little bit weird, though, I agree, hence the debate.

> Option 2: Refactor FileScanTask hierarchy.

I'd be up for refactoring `FileScanTask` even though it will require quite some changes. Any concerns anyone?

Reply:
Refactoring `FileScanTask` sounds good to me too. Maybe as a separate PR before this?

Reply:
There may be a third option. What about parameterizing `Scan` so that it has a task type? Then `DataTableScan` could still use `FileScanTask`, but we could introduce more task types that are different. Then we don't need to change existing scan classes or tasks, but we'd be able to reuse a lot more of the code.

Do you think that would work?

Reply:
That may be reasonable cause we can return `ChangelogScanTask` from this API but I think all other questions still remain. We will have to adapt our readers and refactor the split logic, etc. The main question is whether we consider that worth the effort as technically we can reuse `FileScanTask` for all types of work.

Reply:
Yea I agree, the third option makes sense, but it is a lot of refactoring (not to mention having to migrate external users).  I'm also not sure if it help today's problem of trying to re-use a lot of the reader/split logic done today with FileScanTask, unless we extract some logic there to some new base ScanTask type.